src/Service/Services/MetadataProviders/CosmosSqlMetadataProvider.cs

```diff
@@ -196,27 +195,31 @@ private void ParseSchemaGraphQLFieldsForGraphQLType()
                 {
                     _graphQLTypeToFieldsMap[typeName].Add(field);
                 }
+
+                string modelName = GraphQLNaming.ObjectTypeToEntityName(node);
+                // If the modelName doesn't match, such as they've overridden what's in the config with the directive
+                // add a mapping for the model name as well, since sometimes we lookup via modelName (which is the config name),
+                // sometimes via the GraphQL type name.
+                if (modelName != typeName)
+                {
+                    _graphQLTypeToFieldsMap.TryAdd(modelName, _graphQLTypeToFieldsMap[typeName]);
+                }
             }
         }
 
         public List<string> GetSchemaGraphQLFieldNamesForEntityName(string entityName)
         {
-            List<FieldDefinitionNode>? fields;
-            // Check if entity name has a GraphQL object type name alias. If so, fetch GraphQL object type fields with the alias name
-            foreach (string typeName in _graphQLSingularTypeToEntityNameMap.Keys)
+            if (_graphQLTypeToFieldsMap.TryGetValue(entityName, out List<FieldDefinitionNode>? fields))
```

Comment:
What if the model name was not provided and the entity was aliased using the `singular` type property, the lookup here would fail right? 
The given entityName may not match any graphQLType.

I think we still need the dictionary: _graphQLSingularTypeToEntityNameMap

Reply:
The goal is to try and reduce the need for these maps as they will make it harder when it comes to hot reload so _maybe_, but I'm relying heavily on the test coverage to ensure that we are confident in the changes. We do have test coverage that has some naming mismatches that we relied on the mapping dictionaries for and they are all passing.

Reply:
Im afraid, if we are missing test coverage, this might bite us back. 
Does the naming mismatch tests we have today include case where model name is not provided, and entity is aliased using `singular` property?

Reply:
Does the comment I placed https://github.com/Azure/data-api-builder/pull/1402#discussion_r1244705272 feed into this?

Reply:
No, it doesn't. This comment is about what happens if `@model` isn't provided to the graphql types in the provided GraphQL schema?

All our tests depend on this being true, thats why they don't fail. 
We are missing test coverage to let the engine run successfully even if `@model` isn't specified in the schema for any of the entities (no matter if they are nested or not)
Our current documentation https://learn.microsoft.com/en-us/azure/data-api-builder/get-started/get-started-azure-cosmos-db#add-book-schema-file mentions an incorrect way how to provide model directive. 

I tested the configuration here https://github.com/Azure/data-api-builder/issues/1538#issuecomment-1603043601 for the customer reported issue using their repro schema here(https://github.com/Azure/data-api-builder/issues/1538#issuecomment-1602108032) against 0.7.6 and it works. 
When I checked out your branch and tried to test the same, engine fails to startup with the following stack trace:
```
      Unable to complete runtime initialization. Refer to exception for error details.
      System.InvalidOperationException: Sequence contains no matching element
         at System.Linq.ThrowHelper.ThrowNoMatchException()
         at System.Linq.Enumerable.First[TSource](IEnumerable`1 source, Func`2 predicate)
         at Azure.DataApiBuilder.Service.GraphQLBuilder.GraphQLNaming.ObjectTypeToEntityName(ObjectTypeDefinitionNode node) in D:\hawaii-gql\src\Service.GraphQLBuilder\GraphQLNaming.cs:line 144
         at Azure.DataApiBuilder.Service.Services.MetadataProviders.CosmosSqlMetadataProvider.ParseSchemaGraphQLFieldsForGraphQLType() in D:\hawaii-gql\src\Service\Services\MetadataProviders\CosmosSqlMetadataProvider.cs:line 199
         at Azure.DataApiBuilder.Service.Services.MetadataProviders.CosmosSqlMetadataProvider..ctor(RuntimeConfigProvider runtimeConfigProvider, IFileSystem fileSystem) in D:\hawaii-gql\src\Service\Services\MetadataProviders\CosmosSqlMetadataProvider.cs:line 69
```

This error goes away only after I modify the schema to include `@model(name:"Asset")` directives for ALL the entities. 
Since this has been working with 0.7.6, I think we should continue to support @model to be optional, not rely on its presence to determine the corresponding entityName in the config file.

Reply:
https://github.com/Azure/data-api-builder/pull/1402/commits/66358e54a7ebcdd19388e31da8be42915273df28 should resolve that problem, but I'm going to add a backlog item to expand the test coverage as well so we can ensure it continues to work.

Reply:
Oh, I see https://github.com/Azure/data-api-builder/issues/1556 should cover it.